lib/improver/cli.py

```diff
@@ -0,0 +1,340 @@
+                                            Set(desired_cli_method_args))
+            if len(unmapped_plugin_method_args) > 0:
+                for arg in unmapped_plugin_method_args:
+                    if arg not in actual_plugin_method_defaults[plugin_method]:
+                        raise Exception("Not all plugin methods arguments "
+                                        "were fully specified. Either pass in "
+                                        "from the CLI, or set a default.")
+        # now get the mapping?
+        # certain arguments, e.g: `cube` on the process method, will be created
+        # from the loader and input filepath
+
+        # want some information about the options...
+        plugin_argspecs_dict = {}
+        for plugin_method in CLI.SUPPORTED_PLUGIN_METHODS:
+            method_argspecs = self._derive_argparser_argspecs(
+                                actual_plugin_method_args[plugin_method],
+                                actual_plugin_method_defaults[plugin_method])
+            plugin_argspecs_dict[plugin_method] = method_argspecs
+
+        # we do not get the method args and defaults for LOADER and SAVER,
+        # it is assumed that the loader takes a single filepath, and the saver
+        # takes a cube and a filepath...
+        for special_function, cube_arg_mapping in special_args.items():
+            argspecs_dict[special_function] = []
+            for cube_name, cli_arg in cube_arg_mapping:
+                argspecs_dict[special_function].append([]
+
+        print argspecs_dict
+
+#        # TODO: Need to think about what happens if both have the same flag
+#        # name, because it is wrapped in a list...
+#        plugin_specific_argspecs = init_argspecs + process_argspecs
+#
+#        plugin_description = inspect.getdoc(self.plugin)
+#
+#        cli_args = ArgParser(central_arguments=('input_file', 'output_file'),
+#                             specific_arguments=plugin_specific_argspecs,
+#                             **{'description': plugin_description}
+#                            ).parse_args()
+#
+#        cli_args = vars(cli_args)
+#        self.args = {}
+#        for method_name, method_arg_cli_mapping in cli_method_args.items():
+#            self.args[method_name] = {}
+#            for method_arg, cli_arg in method_arg_cli_mapping:
+#                self.args[method_name][method_arg] = cli_args[cli_arg]
+
+#        self.input_file = cli_args.input_filepath
+#        self.output_file = cli_args.output_filepath
+#
+        # map what was passed in from the command line, to the actual
+        # method arguments
+
+
+    def _check_plugin_conforms(self):
+        """Check that the plugin has the methods which are required for the CLI
+        to function.
+        """
+        # eventually once all plugins inherit from a BasePlugin class, we can
+        # guarantee that they have these methods - for now, manually check
```

Comment:
This will allow us to answer the question: what common functionality do we have. 
An issue?